### ISSUE TITLE
[BUG] cudf.to_orc writes incorrect column and stripe metadata

### ISSUE THREAD

**Issue by randerzander** [rapidsai/cudf#5707]
After writing ORC files with cudf, I'm unable to query them using [Hive](http://hive.apache.org/) on [Tez](http://tez.apache.org/): 

Create a test file:
```
import cudf

df = cudf.DataFrame()
df['id'] = [0, 1, 2, 3]
df['val'] = [1, 2, 3, 4]

df.to_orc('/data/cudf_orc/test.orc')
```

Try to execute a basic query using Hive on Tez ([dockerized setup here](https://github.com/randerzander/docker-hive)):
```
hive> create external table cudf_orc (id bigint, val bigint) stored as orc location '/data/cudf_orc/';
hive> OK
Time taken: 0.064 seconds
hive> select * from cudf_orc;
OK
Failed with exception java.io.IOException:java.lang.RuntimeException: ORC split generation failed with exception: java.lang.IndexOutOfBoundsException: Index: 0
Time taken: 0.1 seconds
```
I get the same behavior if I define the column types as `int`s instead.

Interestingly, Spark can read cudf generated ORC files fine:
```
from pyspark.sql import SparkSession
spark = SparkSession.builder.appName('app').getOrCreate()

spark.read.orc('/data/cudf_orc/').collect()
[Row(id=0, val=1), Row(id=1, val=2), Row(id=2, val=3), Row(id=3, val=4)]
```

Inspection of this cudf generated ORC file using [ORC's CLI tools](https://orc.apache.org/docs/cpp-tools.html#orc-statistics) suggests that cudf doesn't properly write metadata for columns or for stripe stats:
```
~/shared/projects/docker-hive/orc-1.6.3/tools/src$ ./orc-statistics /data/cudf_orc/test.orc 
File /home/rgelhausen/shared/notebooks/cudf_orc/test.orc has 0 columns
File /home/rgelhausen/shared/notebooks/cudf_orc/test.orc has 1 stripes
File /home/rgelhausen/shared/notebooks/cudf_orc/test.orc doesn't have stripe statistics
```

Whereas if I generated the ORC file with Spark, I get the expected metadata.
```
spark.read.orc('/data/cudf_orc/').write.orc('/data/spark_orc')
...

~/shared/projects/docker-hive/orc-1.6.3/tools/src$ ./orc-statistics /data/spark_orc/part-00000-3134c48c-ec95-4df3-92fa-43baf0c0127c-c000.snappy.orc 
File /data/spark_orc/part-00000-3134c48c-ec95-4df3-92fa-43baf0c0127c-c000.snappy.orc has 3 columns
*** Column 0 ***
Column has 4 values and has null value: no

*** Column 1 ***
Data type: Integer
Values: 4
Has null: no
Minimum: 0
Maximum: 3
Sum: 6

*** Column 2 ***
Data type: Integer
Values: 4
Has null: no
Minimum: 1
Maximum: 4
Sum: 10

File /data/spark_orc/part-00000-3134c48c-ec95-4df3-92fa-43baf0c0127c-c000.snappy.orc has 1 stripes
*** Stripe 0 ***

--- Column 0 ---
Column has 4 values and has null value: no

--- Column 1 ---
Data type: Integer
Values: 4
Has null: no
Minimum: 0
Maximum: 3
Sum: 6

--- Column 2 ---
Data type: Integer
Values: 4
Has null: no
Minimum: 1
Maximum: 4
Sum: 10

```

As a way of confirming Hive and my DDL is setup correctly, I _am_ able to read ORC files written by Spark:
```
hive> create external table spark_orc (id bigint, val bigint) stored as orc location '/data/spark_orc/';
OK
Time taken: 0.065 seconds
hive> select * from spark_orc;
OK
0       1
1       2
2       3
3       4
Time taken: 0.127 seconds, Fetched: 4 row(s)
hive> select count(*) from spark_orc;
Query ID = root_20200716182300_dbe12411-d2c1-4012-9c28-9cfe77e0a912
Total jobs = 1
Launching Job 1 out of 1
2020-07-16 18:23:00 Running Dag: dag_1594923772241_0001_2
2020-07-16 18:23:00 Running Dag: dag_1594923772241_0001_2
2020-07-16 18:23:00 Starting to run new task attempt: attempt_1594923772241_0001_2_00_000000_0
2020-07-16 18:23:00 Starting to run new task attempt: attempt_1594923772241_0001_2_00_000000_0
2020-07-16 18:23:00 Completed running task attempt: attempt_1594923772241_0001_2_00_000000_0
2020-07-16 18:23:00 Completed running task attempt: attempt_1594923772241_0001_2_00_000000_0
Status: Running (Executing on YARN cluster with App id application_1594923772241_0001)

----------------------------------------------------------------------------------------------
        VERTICES      MODE        STATUS  TOTAL  COMPLETED  RUNNING  PENDING  FAILED  KILLED  
----------------------------------------------------------------------------------------------
Map 1 .......... container     SUCCEEDED      1          1        0        0       0       0  
Reducer 2        container        INITED      1          0        0        1       0       0  
----------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------
        VERTICES      MODE        STATUS  TOTAL  COMPLETED  RUNNING  PENDING  FAILED  KILLED  
----------------------------------------------------------------------------------------------
Map 1 .......... container     SUCCEEDED      1          1        0        0       0       0  
Reducer 2 ...... container     SUCCEEDED      1          1        0        0       0       0  
----------------------------------------------------------------------------------------------
VERTICES: 02/02  [==========================>>] 100%  ELAPSED TIME: 0.11 s     
----------------------------------------------------------------------------------------------
OK
4
Time taken: 0.476 seconds, Fetched: 1 row(s)
```

I suspect invalid column metadata and missing stripe metadata is the cause of problems reading cudf generated ORC files with Hive.

**Comment by kkraus14** [comment]
@vuule @devavret could one of you tackle this when you get some free cycles? Just assign yourself when you start tackling to make sure not to step on each others toes 😄

**Comment by vuule** [comment]
@randerzander I don't repro the issue when the ORC file is created with `enable_statistics=True`. You should be able to use this to work around the issue for now.
I'll look into whether this behavior is expected with Hive.

**Comment by mythrocks** [comment]
I didn't realize that it was possible to create ORC files without stripe-statistics enabled. :O

Hive requires stripe-statistics if index-based filtering (i.e. predicate pushdown) is enabled, during split-calculation, and in the TableScanOperator during execution.

I suspect the user might be able to get around it if PPD is disabled (i.e. `set hive.optimize.index.filter=false;`), but it really defeats the purpose of using ORC, IMHO.

**Comment by vuule** [comment]
Thank you for clarifying @mythrocks . Based on this, do you think it makes sense to change the `enable_statistics` default to `True`?

**Comment by mythrocks** [comment]
>  do you think it makes sense to change the enable_statistics default to True?

Yes, please. If that is viable, that would be best. 

I think I ran into the same problem, when running tests for CUDF rolling-window-functions. I also found that the ORC files weren't being read. At the time, I was convinced that I'd messed up the data generation. My apologies for not having reported this myself.

**Comment by randerzander** [comment]
Re-opening: after #5825, the snippets in this issue still produce ORC files that can't be read by Hive, and produce the same incorrect output when inspected with ORC's `orc-statistics` CLI tool:

```
./orc-statistics ~/shared/notebooks/cudf_orc/1.orc       │
File /home/rgelhausen/shared/notebooks/cudf_orc/1.orc has 0 columns                                                                                                   │
File /home/rgelhausen/shared/notebooks/cudf_orc/1.orc has 1 stripes                                                                                                   │
File /home/rgelhausen/shared/notebooks/cudf_orc/1.orc doesn't have stripe statistics         
```

**Comment by randerzander** [comment]
Oops, had a conda environment problem with my re-test. This is now resolved =)
